library/common/extensions/filters/http/socket_tag/filter.cc

```diff
@@ -0,0 +1,44 @@
+#include "library/common/extensions/filters/http/socket_tag/filter.h"
+
+#include "envoy/server/filter_config.h"
+
+#include "library/common/network/socket_tag_socket_option_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace SocketTag {
+
+Http::FilterHeadersStatus SocketTagFilter::decodeHeaders(Http::RequestHeaderMap& request_headers,
+                                                         bool) {
+  auto socket_tag_header = Http::LowerCaseString("socket-tag");
+  if (request_headers.get(socket_tag_header).empty()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // The socket-tag header must contain a pair of number separated by a comma, e.g.:
+  // socket-tag: 123,456
+  // The first number contains the UID and the second contains the traffic stats tag.
+  std::string tag_string(request_headers.get(socket_tag_header)[0]->value().getStringView());
+  std::pair<std::string, std::string> data = absl::StrSplit(tag_string, ',');
+  uid_t uid;
+  uint32_t traffic_stats_tag;
+  if (!absl::SimpleAtoi(data.first, &uid) || !absl::SimpleAtoi(data.second, &traffic_stats_tag)) {
+    return Http::FilterHeadersStatus::StopIteration;
```

Comment:
StopIteration or sendLocalReply with error?  I'd think we wanted the latter?  do we have an e2e test for this?

Reply:
Oh but this is a failure case based on getting bad headers from the API - you should totally be able to test sending an invalid request with the filter configured, and getting the local response you expect, so let's do that!

you could also optionally add a test for the happy path, which would create the custom transport socket and make sure hashing works, if we don't have that tested upstream.  even if the tagging is a no-op it would at least cover that the memory model works.  

also clean up the commented out return value above this line?

Reply:
Ah, ok. Java tests added and removed the commented out code.

I looked at  test/common/integration/client_integration_test.cc, which we discussed offline. I wasn't quite sure how to configure it to add the socket tagging support. Maybe we can discuss that offine?

Reply:
Thank you. I am happy to help with the setup of the demo app if you run into any issues with it. It's also possible that the documentation is lacking or making some assumptions - let me know and I am happy to help.

Re experimental app - I think that you can go even one step further (sorry for not pointing this out in my previous comment) and adding a `socket-tag` header to requests performed by the experimental app. This way you would get an end-to-end confirmation that requests with socket-tag go through (we verify whether our experimental example is getting 200 from the server). You could add the header somewhere in [here](https://github.com/envoyproxy/envoy-mobile/pull/2423/files#diff-b15ec59c29644d8c6089ad0ea7370859a690563f75d02c819679270b61ab9520L116)